In [2]:
from anomalib.data import MVTec
from anomalib.models import Patchcore, EfficientAd
from anomalib.engine import Engine

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
def train_and_evaluate(model_class, categories):
    results = {}
    for category in categories:
        print(f"Training and evaluating {model_class.__name__} on {category}")
        datamodule = MVTec(category=category, train_batch_size=1)
        model = model_class()
        engine = Engine(max_epochs=5)
        engine.fit(datamodule=datamodule, model=model, ckpt_path="results/EfficientAd/MVTec/grid/v0/weights/model.ckpt")
        scores = engine.test(datamodule=datamodule, model=model)
        print(scores)
        results[category] = scores
    
    return results, model

categories = ["tile", "leather", "grid"]

In [ ]:
# Patchcore
patchcore_results, patchcore_model = train_and_evaluate(Patchcore, categories)

Training and evaluating Patchcore on tile


GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/core/optimizer.py:182: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer

  | Name                  | Type                     | Params | Mode 
---------------------------------------------------------------------------
0 | model                 | PatchcoreModel           | 24.9 M | train
1 | _transform            | Compose                  | 0      | train
2 | normalization_metrics | MetricCollection         | 0      | train
3 | image_threshol

Epoch 0:   0%|          | 0/230 [00:00<?, ?it/s] 

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Epoch 0: 100%|██████████| 230/230 [00:28<00:00,  8.14it/s]

Python(31524) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31525) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31526) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31527) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31528) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31529) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31530) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(31531) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.



/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp

Epoch 0: 100%|██████████| 230/230 [05:56<00:00,  0.64it/s, pixel_AUROC=0.500, pixel_F1Score=0.153]

`Trainer.fit` stopped: `max_epochs=1` reached.


Epoch 0: 100%|██████████| 230/230 [05:57<00:00,  0.64it/s, pixel_AUROC=0.500, pixel_F1Score=0.153]


F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:419: Consider setting `persistent_workers=True` in 'test_dataloader' to speed up the dataloader worker initialization.
Python(32132) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32133) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32134) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32135) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32136) MallocStackLogging: can't turn off malloc s

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 4/4 [01:19<00:00,  0.05it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.8585858345031738     │
│       image_F1Score       │    0.8571428656578064     │
│        pixel_AUROC        │    0.7486566305160522     │
│       pixel_F1Score       │    0.39787593483924866    │
└───────────────────────────┴───────────────────────────┘

[{'pixel_AUROC': 0.7486566305160522, 'pixel_F1Score': 0.39787593483924866, 'image_AUROC': 0.8585858345031738, 'image_F1Score': 0.8571428656578064}]
Training and evaluating Patchcore on leather


GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/core/optimizer.py:182: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer

  | Name                  | Type                     | Params | Mode 
---------------------------------------------------------------------------
0 | model                 | PatchcoreModel           | 24.9 M | train
1 | _transform            | Compose                  | 0      | train
2 | normalization_metrics | MetricCollection         | 0      | train
3 | image_threshol

Epoch 0:   0%|          | 0/245 [00:00<?, ?it/s] 

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Epoch 0: 100%|██████████| 245/245 [00:10<00:00, 22.79it/s]

Python(32409) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32410) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32411) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32412) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32413) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32414) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32415) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(32416) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.



/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp

Epoch 0: 100%|██████████| 245/245 [07:07<00:00,  0.57it/s, pixel_AUROC=0.500, pixel_F1Score=0.0163]

`Trainer.fit` stopped: `max_epochs=1` reached.


Epoch 0: 100%|██████████| 245/245 [07:08<00:00,  0.57it/s, pixel_AUROC=0.500, pixel_F1Score=0.0163]


F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:419: Consider setting `persistent_workers=True` in 'test_dataloader' to speed up the dataloader worker initialization.
Python(33238) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33239) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33240) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33241) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33242) MallocStackLogging: can't turn off malloc s

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 4/4 [00:25<00:00,  0.16it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │            1.0            │
│       image_F1Score       │     0.994535505771637     │
│        pixel_AUROC        │    0.9899079203605652     │
│       pixel_F1Score       │    0.44024553894996643    │
└───────────────────────────┴───────────────────────────┘

[{'pixel_AUROC': 0.9899079203605652, 'pixel_F1Score': 0.44024553894996643, 'image_AUROC': 1.0, 'image_F1Score': 0.994535505771637}]
Training and evaluating Patchcore on grid


GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/core/optimizer.py:182: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer

  | Name                  | Type                     | Params | Mode 
---------------------------------------------------------------------------
0 | model                 | PatchcoreModel           | 24.9 M | train
1 | _transform            | Compose                  | 0      | train
2 | normalization_metrics | MetricCollection         | 0      | train
3 | image_threshol

Epoch 0:   0%|          | 0/264 [00:00<?, ?it/s] 

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Epoch 0: 100%|██████████| 264/264 [00:13<00:00, 20.23it/s]

Python(33398) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33399) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33400) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33401) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33402) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33403) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33404) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(33405) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.



/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp

Epoch 0: 100%|██████████| 264/264 [09:00<00:00,  0.49it/s, pixel_AUROC=0.526, pixel_F1Score=0.0165]

`Trainer.fit` stopped: `max_epochs=1` reached.


Epoch 0: 100%|██████████| 264/264 [09:00<00:00,  0.49it/s, pixel_AUROC=0.526, pixel_F1Score=0.0165]


F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:419: Consider setting `persistent_workers=True` in 'test_dataloader' to speed up the dataloader worker initialization.
Python(34284) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(34285) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(34286) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(34287) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(34288) MallocStackLogging: can't turn off malloc s

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 3/3 [00:27<00:00,  0.11it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.9766082167625427     │
│       image_F1Score       │    0.9642857313156128     │
│        pixel_AUROC        │    0.9809244275093079     │
│       pixel_F1Score       │    0.37669840455055237    │
└───────────────────────────┴───────────────────────────┘

[{'pixel_AUROC': 0.9809244275093079, 'pixel_F1Score': 0.37669840455055237, 'image_AUROC': 0.9766082167625427, 'image_F1Score': 0.9642857313156128}]


In [ ]:
# EfficientAD
efficientAd_results, efficientAd_model = train_and_evaluate(EfficientAd, categories)

GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead
F1Score class exists for backwards compatibility. It will be removed in v1.1. Please use BinaryF1Score from torchmetrics instead


Training and evaluating EfficientAd on tile



  | Name                  | Type                     | Params | Mode 
---------------------------------------------------------------------------
0 | model                 | EfficientAdModel         | 8.1 M  | train
1 | _transform            | Compose                  | 0      | train
2 | normalization_metrics | MetricCollection         | 0      | train
3 | image_threshold       | F1AdaptiveThreshold      | 0      | train
4 | pixel_threshold       | F1AdaptiveThreshold      | 0      | train
5 | image_metrics         | AnomalibMetricCollection | 0      | train
6 | pixel_metrics         | AnomalibMetricCollection | 0      | train
---------------------------------------------------------------------------
8.1 M     Trainable params
0         Non-trainable params
8.1 M     Total params
32.235    Total estimated model params size (MB)
46        Modules in train mode
7         Modules in eval mode
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/

Training: |          | 0/? [00:00<?, ?it/s]

Python(37629) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37630) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37631) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37632) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37633) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37634) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37635) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Python(37636) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated

Epoch 0:   0%|          | 0/230 [00:00<?, ?it/s] 


/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/core/module.py:516: You called `self.log('train_st', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/core/module.py:516: You called `self.log('train_ae', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/lightning/pytorch/core/module.py:516: You called `self.log('train_stae', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python

Epoch 0:  66%|██████▌   | 152/230 [01:05<00:33,  2.33it/s, train_st_step=10.90, train_ae_step=0.874, train_stae_step=0.000494, train_loss_step=11.70]


Detected KeyboardInterrupt, attempting graceful shutdown ...


In [16]:
print('----------------------------------')

print("PatchCore Results")
for key, value in patchcore_results.items():
    print(key, " : ", value[0]['image_AUROC'])

print('----------------------------------')

print("EfficientAD Results")
for key, value in efficientAd_results.items():
    print(key, " : ", value[0]['image_AUROC'])

print('----------------------------------')

----------------------------------
PatchCore Results
tile  :  0.8585858345031738
leather  :  1.0
grid  :  0.9766082167625427
----------------------------------
EfficientAD Results
tile  :  0.9635642766952515
leather  :  0.879755437374115
grid  :  0.9807853698730469
----------------------------------


### **AUROC Scores**

| **Category** | **PatchCore**           | **EfficientAD**        |
| ------------ | ----------------------- | ---------------------- |
| tile         | 0.8585858345031738      | 0.9635642766952515     |
| leather      | 1.0                     | 0.879755437374115      |
| grid         | 0.9766082167625427      | 0.9807853698730469     |


In [ ]:
from anomalib.engine import Engine
from anomalib.models import Patchcore, EfficientAd

# Loading models from the checkpoint

checkpoints = {
    "patch_core" : {
        "tile" : "results/Patchcore/MVTec/tile/v8/weights/lightning/model.ckpt",
        "leather" : "results/Patchcore/MVTec/leather/v2/weights/lightning/model.ckpt",
        "grid" : "results/Patchcore/MVTec/grid/v2/weights/lightning/model.ckpt"
    },
    "efficient_ad" : {
        "tile" : "results/EfficientAd/MVTec/tile/v2/weights/lightning/model.ckpt",
        "leather" : "results/EfficientAd/MVTec/leather/v0/weights/lightning/model.ckpt",
        "grid" : "results/EfficientAd/MVTec/grid/v0/weights/lightning/model.ckpt"
    }
}

models = {
    "patch_core" : Patchcore(),
    "efficient_ad" : EfficientAd()
}

In [53]:
engine = Engine(max_epochs=5)

models = {
    "patch_core" : Patchcore(),
    "efficient_ad" : EfficientAd()
}

categories = ["tile", "leather", "grid"]

datamodule = {}
scores = {}

for model in models:
    for category in categories:
        datamodule.setdefault(model, {})[category] = MVTec(category=category, train_batch_size=1)
        scores.setdefault(model, {})[category] = engine.test(datamodule=datamodule[model][category], model=models[model], ckpt_path=checkpoints[model][category])

GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/anomalib/engine/engine.py:391: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case wh

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 4/4 [00:42<00:00,  0.09it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.8585858345031738     │
│       image_F1Score       │    0.8571428656578064     │
│        pixel_AUROC        │    0.7486566305160522     │
│       pixel_F1Score       │    0.39787593483924866    │
└───────────────────────────┴───────────────────────────┘

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/anomalib/engine/engine.py:391: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this e

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 4/4 [00:39<00:00,  0.10it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │            1.0            │
│       image_F1Score       │     0.994535505771637     │
│        pixel_AUROC        │    0.9899079203605652     │
│       pixel_F1Score       │    0.44024553894996643    │
└───────────────────────────┴───────────────────────────┘

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/anomalib/engine/engine.py:391: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this e

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 3/3 [00:17<00:00,  0.17it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.9766082167625427     │
│       image_F1Score       │    0.9642857313156128     │
│        pixel_AUROC        │    0.9809244275093079     │
│       pixel_F1Score       │    0.37669840455055237    │
└───────────────────────────┴───────────────────────────┘

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/anomalib/engine/engine.py:391: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this e

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 4/4 [00:36<00:00,  0.11it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.9635642766952515     │
│       image_F1Score       │    0.9268292784690857     │
│        pixel_AUROC        │    0.8119826316833496     │
│       pixel_F1Score       │    0.5696297287940979     │
└───────────────────────────┴───────────────────────────┘

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/anomalib/engine/engine.py:391: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this e

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 4/4 [00:25<00:00,  0.15it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │     0.879755437374115     │
│       image_F1Score       │    0.8863636255264282     │
│        pixel_AUROC        │    0.9580346345901489     │
│       pixel_F1Score       │    0.4297158718109131     │
└───────────────────────────┴───────────────────────────┘

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/anomalib/engine/engine.py:391: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this e

Testing: |          | 0/? [00:00<?, ?it/s]

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.

Testing DataLoader 0: 100%|██████████| 3/3 [00:44<00:00,  0.07it/s]


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.9807853698730469     │
│       image_F1Score       │    0.9649122953414917     │
│        pixel_AUROC        │     0.907240629196167     │
│       pixel_F1Score       │    0.4664262533187866     │
└───────────────────────────┴───────────────────────────┘

In [54]:
for model in models:
    for key, value in scores[model].items():
        print(key, " : ", value[0]["image_AUROC"])

tile  :  0.8585858345031738
leather  :  1.0
grid  :  0.9766082167625427
tile  :  0.9635642766952515
leather  :  0.879755437374115
grid  :  0.9807853698730469


#### Initialize Qdrant, Extract features and stores in Qdrant db

In [ ]:
import numpy as np
import torch.nn.functional as F
from sklearn.decomposition import PCA
from qdrant_client import QdrantClient
from qdrant_client.models import PointStruct
from qdrant_client.http import models as qdrant_models


client = QdrantClient(
    url="https://0c35ee0e-5843-4fbc-92e7-ad3e87a8ae43.us-east4-0.gcp.cloud.qdrant.io:6333", 
    api_key="RdYLvFgkSCBmF6VEdUr47ThCwyNKr5HCOy4YjujdWVA4wYsPWVnPiQ",
    timeout=300
)

collection_name = 'anomaly_features'

client.recreate_collection(
    collection_name=collection_name,
    vectors_config=qdrant_models.VectorParams(size=39200, distance=qdrant_models.Distance.COSINE)
)

def add_to_qdrant(image_id, feature):
    print(len(feature.tolist()))
    point = PointStruct(
        id=image_id,
        vector=feature.tolist()
    )
    client.upsert(
        collection_name=collection_name,
        points=[point]
    )

pca = PCA(n_components=50)

def extract_features(model, train_loader):
    features = {}
    for idx, batch in enumerate(iter(train_loader)):
        images = batch['image']
        image_id = batch['image_path'][0].split('/')[-1].split('.')[0]
        embeddings = model(images)
        print(embeddings.shape)
        reduced_embeddings = pca.fit_transform(embeddings)
        print(reduced_embeddings.shape)
        feature = reduced_embeddings.reshape(-1)
        print(type(feature), feature.shape)
        add_to_qdrant(image_id=idx, feature=feature)
        features[idx] = (image_id, feature)
        print(idx)
    return features

train_loader = datamodule["patch_core"]["tile"].train_dataloader()
features = extract_features(models["patch_core"], train_loader)

/var/folders/mt/vkl4jzfs66jcxhsr9_05n3lm0000gn/T/ipykernel_12539/880236198.py:21: DeprecationWarning: `recreate_collection` method is deprecated and will be removed in the future. Use `collection_exists` to check collection existence and `create_collection` instead.
  client.recreate_collection(
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/.venv/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
/Users/srik

torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
0
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
1
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
2
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
3
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
4
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
5
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
6
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
7
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
8
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
9
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
10
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
11
torch.Size([784, 1536])
(784, 50)
<class 'numpy.ndarray'> (39200,)
39200
12
torch.Size([784, 1536]

#### Selecting a image as query for similarity search

In [73]:
data = datamodule["patch_core"]["tile"]

data.setup()

train_loader = data.train_dataloader()
train_dataset = train_loader.dataset
query_image = train_dataset[10]['image'].unsqueeze(0)
query_image_path = train_dataset[10]['image_path']
print(query_image.shape)

torch.Size([1, 3, 224, 224])


#### Extracting the feature for the query image

In [74]:
import numpy as np
import torch.nn.functional as F
from sklearn.decomposition import PCA
from qdrant_client import QdrantClient
from qdrant_client.models import PointStruct
from qdrant_client.http import models as qdrant_models

collection_name = 'anomaly_features'

pca = PCA(n_components=50)

def extract_features(model, image, image_path):
    print(image_path)
    image_id = image_path.split('/')[-1].split('.')[0]
    print(image.shape)
    embeddings = model(image)
    print(embeddings.shape)
    reduced_embeddings = pca.fit_transform(embeddings)
    feature = reduced_embeddings.reshape(-1)
    return feature, image_id

query_feature, image_id = extract_features(models["patch_core"], image=query_image, image_path=query_image_path)
print(query_feature.shape)

/Users/srikrishna/Desktop/Semester 1/Artificial Intelligence/cs-gy-6613-assignments/Assignment 2b/datasets/MVTec/tile/train/good/010.png
torch.Size([1, 3, 224, 224])
torch.Size([784, 1536])
(39200,)


In [75]:
def similarity_search(image_features, top_k=5):
    result = client.search(
        collection_name=collection_name,
        query_vector=image_features,  # Feature vector of the input image
        limit=top_k  # Number of similar images to return
    )
    return result

# Example usage: Assuming `new_image_features` is the feature vector of a new image
results = similarity_search(query_feature)
print(results)

[ScoredPoint(id=175, version=175, score=0.99987054, payload={}, vector=None, shard_key=None, order_value=None), ScoredPoint(id=95, version=95, score=0.38834572, payload={}, vector=None, shard_key=None, order_value=None), ScoredPoint(id=113, version=113, score=0.29995775, payload={}, vector=None, shard_key=None, order_value=None), ScoredPoint(id=107, version=107, score=0.2936305, payload={}, vector=None, shard_key=None, order_value=None), ScoredPoint(id=152, version=152, score=0.29232284, payload={}, vector=None, shard_key=None, order_value=None)]


In [89]:
folder_path = "./datasets/MVTec/tile/train/good/"
for result in results:
    index = int(features[result.id][0])
    image = train_dataset[index]["image"]
    image_path = folder_path + features[result.id][0] + ".png"
    print(image_path)


./datasets/MVTec/tile/train/good/010.png
./datasets/MVTec/tile/train/good/157.png
./datasets/MVTec/tile/train/good/193.png
./datasets/MVTec/tile/train/good/196.png
./datasets/MVTec/tile/train/good/142.png
